src/main/java/com/google/devtools/build/lib/bazel/repository/cache/RepoContentsCache.java

```diff
@@ -39,4 +44,38 @@ public Path contentsDir(String predeclaredInputHash) {
     Preconditions.checkState(path != null);
     return path.getRelative(predeclaredInputHash);
   }
+
+  private Path lockFile(String predeclaredInputHash) {
+    Preconditions.checkState(path != null);
+    return path.getRelative(predeclaredInputHash + ".lock");
+  }
+
+  public FileSystemLock lockShared(String predeclaredInputHash) throws IOException {
+    return FileSystemLock.getShared(lockFile(predeclaredInputHash));
+  }
+
+  /** Moves a freshly fetched repo into the contents cache. */
+  public void moveToCache(
+      Path fetchedRepoDir, Path fetchedRepoMarkerFile, String predeclaredInputHash)
+      throws IOException {
+    Preconditions.checkState(path != null);
+
+    Path cacheRepoDir = contentsDir(predeclaredInputHash);
+    Path cacheRecordedInputsFile = recordedInputsFile(predeclaredInputHash);
+
+    try (FileSystemLock l = FileSystemLock.getExclusive(lockFile(predeclaredInputHash))) {
+      cacheRepoDir.deleteTree();
+      cacheRepoDir.createDirectoryAndParents();
+      // Move the fetched marker file to a temp location, so that if following operations fail, both
+      // the fetched repo and the cache locations are considered out-of-date.
+      Path tempMarker = path.getRelative(predeclaredInputHash + ".temp_recorded_inputs");
+      FileSystemUtils.moveFile(fetchedRepoMarkerFile, tempMarker);
+      // Now perform the move, and afterwards, restore the marker file.
+      FileSystemUtils.moveTreesBelow(fetchedRepoDir, cacheRepoDir);
```

Comment:
Could we optimistically try a `renameTo` first before falling back to moving the subtrees? `moveTreesBelow` appears to move all files individually, which could be much slower.

Reply:
Done.